### PR TITLE
Candis docker service separated into development and production 

### PR DESCRIPTION
**Pull request by promehul**
###### Explanation About What Code Achieves:

  -   This code dockerizes the candis app completely . Now with the help of only two commands anyone can setup the candis app on their local system and start contributing to our app. Apart from development, the dockerization has been divided for production as well. 
 - Requirements files has been divided seperately for development and production with development dependencies including both the requirements file.
 - Then two seperate dockerfiles has been written seperately for dev and prod . Development docker file uses the concept of volumes so that the changes done in the local are reflected back in the container as well. On the other hand production dockerfile uses copy the candis code into the container to ensure immutability and reproducibility.
- Then I seperated the docker-compose files for both . 

###### Screenshots/GIFs:

  - Screenshot

###### Steps To Test:

  -  Use `docker-compose -f development.yml build` to build the docker image for development and then use 
        `docker-compose -f development.yml up` to start the container.
 - Use `docker-compose -f production.yml build` to build the docker image for production and then use 
        `docker-compose -f production.yml up` to start the container for the production stage.
###### TODOs:

  - None

### CODE REVIEW

File: compose/development/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    && add-apt-repository 'deb [arch=amd64,i386] https://cran.rstudio.com/bin/linux/ubuntu xenial/' \
+    && apt-get update \
+    && apt-get install -y r-base 
+
+COPY ./requirements /requirements
+
+RUN pip3 install --upgrade pip
+RUN pip3 install setuptools wheel \
+    && pip3 install numpy pyyaml \
+    && pip3 install --no-binary javabridge -r /requirements/development.txt  -r /requirements/production.txt
```
Comment (rupav) [inline compose/development/Dockerfile]: add all dependencies in development.txt, production.txt will be the subset of it.

Comment (promehul) [reply compose/development/Dockerfile]: @rupav  that has one issue. Any new dependency then has to be added in both the files.
In this only one file u need to add.

Comment (rupav) [reply compose/development/Dockerfile]: Yes, But development.txt implies that all dependencies in it should be sufficient to start on local setup for new developers. Not everyone will directly jump into docker for setup.

Comment (promehul) [reply compose/development/Dockerfile]: True. Will do it . But I don't think  if we have dockerized the whole app , then anyone will try installing candis by installing each dependency individually

**Comment by promehul** [comment]
@rupav  changes done !! Any more reviews? I will push all the changes together.

File: compose/production/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+RUN pip3 install --upgrade pip
+RUN pip3 install setuptools wheel \
+    && pip3 install numpy pyyaml \
+    && pip3 install --no-binary javabridge -r ./requirements/production.txt
+
+ENV PYTHONPATH="/app/candis"
+
+EXPOSE 5000
+
+CMD ["python3", "-m", "candis"]
```
Comment (achillesrasquinha) [inline compose/production/Dockerfile:37]: This apparently uses the dev server. Try using gunicorn instead.

Comment (achillesrasquinha) [reply compose/production/Dockerfile:37]: or uwsgi

Comment (achillesrasquinha) [reply compose/production/Dockerfile:37]: create a /start script or an /entrypoint script

File: development.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+x-defaults: &defaults
+  restart: unless-stopped
+
+services:
+  db:
+    <<: *defaults
+    image: postgres
+    networks:
+      - db
+    expose:
```
Comment (achillesrasquinha) [inline development.yml:12]: nope.

File: development.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    depends_on:
+      - db
+    volumes: 
+      - .:/app    
+  cache:
+    <<: *defaults
+    image: redis
+    networks:
+      - cache
+  proxy:
```
Comment (achillesrasquinha) [inline development.yml:36]: why use nginx for development?

File: development.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    expose:
+      - "5432"
+  app:
+    <<: *defaults
+    build: 
+      context: .
+      dockerfile: ./compose/development/Dockerfile
+    environment:
+      CANDIS_DATABASE_HOST: db
+      REDIS_HOST: cache
```
Comment (achillesrasquinha) [inline development.yml:21]: expose ports

File: production.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+x-defaults: &defaults
+  restart: unless-stopped
+
+services:
+  db:
+    <<: *defaults
+    image: postgres
+    networks:
+      - db
+    expose:
```
Comment (achillesrasquinha) [inline production.yml:12]: nope.

File: production.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      - proxy
+
+    depends_on:
+      - db
+  cache:
+    <<: *defaults
+    image: redis
+    networks:
+      - cache
+  proxy:
```
Comment (achillesrasquinha) [inline production.yml:34]: you might also need to add certbot

**Comment by promehul** [comment]
@achillesrasquinha and @rupav  any follow up?

File: compose/production/Dockerfile (hunk 2 above)
Comment (achillesrasquinha) [inline compose/production/Dockerfile:37]: @promehul

**Review by achillesrasquinha — CHANGES_REQUESTED** [review]

**Pull request closed without merging**